client/asset/eth/eth_test.go

```diff
@@ -2773,6 +2787,449 @@ func TestLocktimeExpired(t *testing.T) {
+	_, eth, node, shutdown := tassetWallet(assetID)
+	defer shutdown()
+
+	var secretHash [32]byte
+	copy(secretHash[:], encode.RandomBytes(32))
+	header := &types.Header{Number: big.NewInt(200)}
+	baseCtx := context.Background()
+
+	tests := []struct {
+		name            string
+		bestHeader      *types.Header
+		bestHeaderErr   error
+		coinID          dex.Bytes
+		requestResult   error
+		contractVersion uint32
+		secretHash      [32]byte
+		cancelCtx       bool
+		invalidContract bool
+		duplicate       bool
+
+		expectRequestQueued bool
+		expectError         bool
+	}{
+		{
+			name:                "ok",
+			bestHeader:          header,
+			coinID:              encode.RandomBytes(32),
+			contractVersion:     1,
+			secretHash:          secretHash,
+			expectRequestQueued: true,
+		},
+		{
+			name:                "context cancelled",
+			bestHeader:          header,
+			coinID:              encode.RandomBytes(32),
+			contractVersion:     1,
+			secretHash:          secretHash,
+			cancelCtx:           true,
+			expectRequestQueued: true,
+			expectError:         true,
+		},
+		{
+			name:                "invalid coin id",
+			bestHeader:          header,
+			coinID:              encode.RandomBytes(31),
+			contractVersion:     1,
+			secretHash:          secretHash,
+			expectRequestQueued: false,
+			expectError:         true,
+		},
+		{
+			name:                "invalid contract",
+			bestHeader:          header,
+			coinID:              encode.RandomBytes(32),
+			contractVersion:     1,
+			secretHash:          secretHash,
+			invalidContract:     true,
+			expectRequestQueued: false,
+			expectError:         true,
+		},
+		{
+			name:                "best header err",
+			bestHeader:          header,
+			coinID:              encode.RandomBytes(32),
+			contractVersion:     1,
+			secretHash:          secretHash,
+			bestHeaderErr:       errors.New("best header err"),
+			expectRequestQueued: false,
+			expectError:         true,
+		},
+		{
+			name:                "duplicate request",
+			bestHeader:          header,
+			coinID:              encode.RandomBytes(32),
+			contractVersion:     1,
+			secretHash:          secretHash,
+			duplicate:           true,
+			expectRequestQueued: false,
+			expectError:         true,
+		},
+	}
+	for _, test := range tests {
+		ctx, cancel := context.WithCancel(baseCtx)
+		defer cancel()
+
+		node.bestHdrErr = test.bestHeaderErr
+		node.bestHdr = test.bestHeader
+
+		var txHash [32]byte
+		copy(txHash[:], test.coinID)
+
+		if test.duplicate {
+			eth.confirmRedemptionReqs[txHash] = &confirmRedemptionRequest{}
+		}
+
+		var contract dex.Bytes
+		if test.invalidContract {
+			contract = encode.RandomBytes(30)
+		} else {
+			contract = dexeth.EncodeContractData(test.contractVersion, test.secretHash)
+		}
+
+		var wg sync.WaitGroup
+		wg.Add(1)
+		var err error
+		go func() {
+			err = eth.ConfirmRedemption(ctx, test.coinID, contract)
+			wg.Done()
+		}()
+
+		time.Sleep(10 * time.Millisecond)
```

Comment:
Here I'm just waiting for the `confirmRedemptionRequest` to be added to the map to be able to check that the request was created correctly. I guess I could combine this with `TestCheckConfirmRedemptions` and not check that the request is populated correctly.

Reply:
How do you feel about the change? It still uses `time.Sleep` but avoids any false negatives.

Reply:
I'm not crazy about Sleep in test either, but usually when we hack into internals or make complex signaling mechanisms in our stubs, the test become difficult to maintain.

So I suggest avoiding sleeps unless doing so relies heavily on implemention details that are not part of the API's documented contact with the consumer.  What happens if we change some internal thing, like some otherwise unimportant order of operations, and the test was start failing even though the exported function is really working the same